### ISSUE TITLE
Add real time player behavior analysis

### ISSUE THREAD

**Issue by ETLegacyTeam** [etlegacy/etlegacy#1021]
Author: @rmarquis (Spyhawk)
Date: 2017-03-13
Redmine Issue: 1021, https://dev.etlegacy.com/issues/1021

---

I have trouble calling this an "anticheat" since it's radically different from traditional anticheat, so let's say this feature is about an "automated monitoring of players behavior" using Machine Learning techniques.
The necessary framework exists already and minimal work is required to implement its interface in ETL code. The core feature itself which is/will be implemented in python, leveraging the power of the scikit-learn library.

This is the perfect companion of the server side anti-wallhack feature (by Laszlo) that we already implemented.

> I worked on an opensource anticheat architecture, completely server-sided (so that cheaters can’t avoid it), based on behavior analysis automated by machine learning.

> There are several research papers about automated behavior analysis to detect cheaters, it works well. My architecture goes beyond simply doing behavior analysis by providing a full architecture to produce the necessary behavior data from servers automatically, then ease the analysis (using any algorithm you want), the sharing of behavior analysis parameters (so that server administrators can share the parameters or even the behaviors databases in order to construct big datasets to get better parameters) and finally allows for ban/kick/demo recording/any rcon command as a final measure.

> The whole project is here, made for OpenArena, but easily translatable to any ioq3 based game:
> * https://github.com/lrq3000/openarena_engine_oacs
> * https://github.com/lrq3000/oacs

> That being said, I stopped the project because of a lack of time, but it fully works. The only downfall is that the algorithms I used are not efficient enough. I plan to add support for scikit-learn, leveraging the big library of machine learning algorithms scikit-learn offers. Also, better behavior features could provide better results, such as a better reaction time estimator and a crosshair estimator (is the crosshair aiming at a player’s head? for how long? etc.).

> This approach was implemented by the mod ExcessivePlus (along with an enhanced antiwallhack but stemming from the same Laszlo patch). If anybody is interested into reviving this project, I can supervise the effort and provide the update to support scikit-learn.

> To clarify what is the goal of such a system: with behavior analysis, you analyze what are "normal" players behavior (non-cheaters), and from that you can detect "anomalous" behaviors (cheaters). This has the big advantage that you just need to get data from normal players, which is readily available. It will also be robust enough to detect new cheats (because we don’t model the cheats anyway but the non-cheat).

> In the end, cheats can still work, but only if they do not display "anomalous" behavior, such as a too high precision or unnatural human movements or reaction time. So in the end, such cheats are not a threat anymore, since they do not give a "surnatural" power, because they simply are not allowed to trespass the "natural ability" threshold.

> Of course that’s in theory, and in practice there will be several false positives. But the goal of the system is NOT to auto-ban or auto-kick (even if it is possible), but rather to autodetect suspicious behavior, and launch an autorecord of them, for later review by the server administrator. This would also be the most efficient, as delaying the punition will reduce the clues cheaters can have (indeed if they get kicked instantly, they can iteratively refine their cheats, whereas if the punition is delayed, they can’t know what version of their cheat did that).

> I’m convinced this is the best approach for an opensource anticheating system, as it will leverage the collaborative nature of opensource, because:
> 1. you can share datasets and parameters, everything is anonymized, the ids are not necessary for the behavior analysis).
> 2. having the database/parameters does not provide the program (since it is learnt by a machine), and even if learnt, it cannot be bypassed (since everything is analyzed server-side). This is why this approach is also being investigated by AAA studios currently such as Battlefields.

I'm personally familiar with ML and scikit-learn, and seeing that the existing framework does all the nitty-gritty already, most of the effort could be spent on doing the relevant ML work.

Some references:

* [Open Anti-Cheat system: An opensource anti-cheat system for realtime online multiplayer games](https://www.cse.cuhk.edu.hk/~cslui/PUBLICATION/detect_cheat.pdf) (Augmented Gaussian)
* [Detecting Cheaters for Multiplayer Games: Theory, Design and Implementation](https://www.cse.cuhk.edu.hk/~cslui/PUBLICATION/detect_cheat.pdf) (dynamic Baysian network)
* [US20190291008A1](https://patentswarm.com/patents/US20190291008A1)

**Comment by rafal1137** [comment]
Just want to point out I found more updated reference by far.
Here it is:

- http://www.diva-portal.se/smash/get/diva2:1431282/FULLTEXT01.pdf

**Comment by rmarquis** [comment]
Wow, that's a good one. It's referring to Recurrent Neural Network, which are indeed a more modern solution than the multivariate Gaussian implementation of OACS.
Thanks for posting!

**Comment by rafal1137** [comment]
After a fine lecture, I did decided to spark some ideas here:

- Solution 1: Save all the data of player aiming to a file as he plays a game ofc file has to be sent.

- Solution 2: Let player record a demo and using Demo Viewer by a Admin or a person responsible for reviewing a demos record this data to a file.

- Solution 3: And let me take a shortcut here. By a Server Demo - But is that even possible since you are following a player thus will you recieve his data regarding aiming ? 


Those are about getting that data. Next step would be writing a model, training, validating, and getting final model that we can use for detecting those that are unfair.

PS: If you did read this thesis you can see some similiarities as Solution 1 and 2 are the ones that Author of this thesis had in mind.

**Comment by rmarquis** [comment]
Solution 1 and 3 makes the most sense to me, as they can be done server-side and thus fully automated for real time analysis.

Overall, I'd say the main difficulty isn't even about implementing the algorithm or training it, but handling the data generation. Realistically, we'd only be able to work with a community with plenty of active players (which means TM), and that kind of stuff generates *a lot* of data (probably hundred of MB or GB of it each day), which makes it awkward to work in an iterative manner with a community that actually wants to have fun.

There is a lot of annoying boring work in preparing data until we "get it right", and quite of admin work required that I am not sure everybody would be pleased with. That's not something to overlook.

We'd also need Python and some ML framework installed on the gaming server, so that's another kettle of fish.

**Comment by rafal1137** [comment]
> 
> 
> Solution 1 and 3 makes the most sense to me, as they can be done server-side and thus fully automated for real time analysis.
> 
> Overall, I'd say the main difficulty isn't even about implementing the algorithm or training it, but handling the data generation. Realistically, we'd only be able to work with a community with plenty of active players (which means TM), and that kind of stuff generates _a lot_ of data (probably hundred of MB or GB of it each day), which makes it awkward to work in an iterative manner with a community that actually wants to have fun.
> 
> There is a lot of annoying boring work in preparing data until we "get it right", and quite of admin work required that I am not sure everybody would be pleased with. That's not something to overlook.
> 
> We'd also need Python and some ML framework installed on the gaming server, so that's another kettle of fish.

Yes I like the idea of getting those data by server-side but this will require to server do some work here and there, About hardware requirements would they change here ?
We know the more data the merrier :D but let me honest here this data would be required to train/validate model I do not think we would require it later unless I am mistaken, will ask MrMuppet about how much data their server is generating.

In this thesis they did posted how raw data is looking for them:
> L02/03/2020−13:08:46:[onplayerruncmd.smx] AimingAtClient=false Pitch=5.207493 Yaw=−69.532730

But dunno how our system is working when you are aiming at player or just your crosshair is near, so far I know that when you are aiming at player you are getting the data about it.

After that data collection yes tunning the parameters is kind of boring stuff that have to be done.
If more work would be required for admins yes I get the point on that but would you rather have no way of catching a cheaters and let your server be free zone for any kind of unfair game. But cleaning a server yes would require some work for them which would be the advantage here. Also not every server would run it yes pubs server would be the ones. It would be server admin to decide to work with it or not. Competition Scene - it would be turned on all the time and there is no doubt.
Extra dependencies would be required to be installed if you want to make a use of it and I would be appreaciated if the time it took to install extra software would be paid off with catching a cheaters.

More comments on that one are apprecieated here.

**Comment by rmarquis** [comment]
> this data would be required to train/validate model I do not think we would require it later unless I am mistaken

Yes, we need to collect quite some data for training the model. But once this is done, we *also* need to collect the same data to compare it to the model, as to classify players as cheater or not (we just don't need to keep it and can throw it away after each map, but we need to at least stream it into the python system).

> About hardware requirements would they change here ?

I have no idea. Fundamentally, this isn't much different from the existing logging feature. However, it happens at a much more frequent rate:

> In this thesis they did posted how raw data is looking for them:
> > L02/03/2020−13:08:46:[onplayerruncmd.smx] AimingAtClient=false Pitch=5.207493 Yaw=−69.532730

They also write they do it at 128 Hz, so that data is logged... 128 times a second. If we do it for each player, assuming 30 or 40 players, the log size is becoming quite big quite fast. Not saying we absolutely requires to do it that much, but the main idea is that the higher the frequency we use, the easier it is to distinguish cheats like humanized aimbot.

> But dunno how our system is working when you are aiming at player or just your crosshair is near, so far I know that when you are aiming at player you are getting the data about it.

What they do is kinda basic (data when the cursor is on a target), but a more useful prediction could be done by registering, for example, the relative position of the target to the player (again, for humanized aimbot that include some randomness to hide they aren't human). That's what we call "features selection" in the ML sense , and the more diverse the criteria are the better it is to model cheat detection accurately.

> If more work would be required for admins yes I get the point on that but [...]

Yes, on paper everyone would agree that an anti-cheat is a must have. I'm just saying it is quite demanding for server admins, and on the longer term too. We might find ourselves collecting data, train a model a realize it's not good enough. We might have to adjust ML features selection and throw away everything that's been done prior to that, in an iterative process. It's somewhat different than implementing features in the code, and I'd try to ensure everyone is on the same page and understand what's such a project entails before starting it.

**Comment by rafal1137** [comment]
> What they do is kinda basic (data when the cursor is on a target), but a more useful prediction could be done by registering, for example, the relative position of the target to the player (again, for humanized aimbot that include some randomness to hide they aren't human). That's what we call "features selection" in the ML sense , and the more diverse the criteria are the better it is to model cheat detection accurately.

Yes they had pointed it out about the possibilities of cheat having such more advanced stuff.

> Yes, on paper everyone would agree that an anti-cheat is a must have. I'm just saying it is quite demanding for server admins, and on the longer term too. We might find ourselves collecting data, train a model a realize it's not good enough. We might have to adjust ML features selection and throw away everything that's been done prior to that, in an iterative process. It's somewhat different than implementing features in the code, and I'd try to ensure everyone is on the same page and understand what's such a project entails before starting it.

About this getting everyone on the same page. How are we gonna inform server admins about possibility of creating/using Anti Cheat Feature based on ML with all the advances and disadvantages ?

**Comment by rmarquis** [comment]
> Yes they had pointed it out about the possibilities of cheat having such more advanced stuff.

You might have caught that already, but they're doing also something simpler in that they model only one "good" behaviour and one specific cheat. Obviously, the anomaly detection approach would fit better, and we can use RNN for doing better than Gaussian. All of this requires more data.

> How are we gonna inform server admins about possibility of creating/using Anti Cheat Feature based on ML with all the advances and disadvantages ?

"Excuse me Sir, do you have a moment to talk about our lord and savior, the mighty Anti-Cheat? The path to salvation is long and painful, but what is this compared to eternal salvation?"

Honestly, your guess is as good as mine. They should however be really aware this is *work*, and most probably more work for them than for us. Writing the code is the simple part.

**Comment by OSdevTom** [comment]
Hello, 
I can work on this. I read the material. But I have built real time behavior analysis systems for several organizations which are a complex mix of AI and business rules. 
I believe the easiest path forward is to create a little parser from demo files that would create an ABT(analytical based table) in a timeseries format. I need a bit of help for that as I am not familiar with writing from the engine into the database. 
@rmarquis I think you or @mittermichal could help me with it. 
With the following format and content: 
![image](https://user-images.githubusercontent.com/79510802/145013868-e4c48b1b-2946-4301-9008-b88c451ac5d7.png)
With this I can develop several AI pipelines in python to see if this could work. 
Potentially, it may be difficult for you to calculate interval values (velocity), if it is i can do it in the pipeline but then i need the positions. 
The time interval between observation should be the same as the fastest shooting rate possible in the game. 
When this data pipeline creates a decent ABT we can execute it against a lot of demos.
Then we can test several algorithms. I am a bit worried about RNN as they may require quite some power at inference but let's see. I think if we can have a performant system with demos, it is already something. 
Cheers

**Comment by rmarquis** [comment]
Sorry, demos are one aspect of the code I never quite delved into.

This said, I think starting with demo is a good idea, as the same principle could then applied for real time game. Doing it on demos (and server demos specifically) first might also alleviate some complexity in handling lot of data on live server, as a starting point.

**Comment by HeDo88TH** [comment]
I think that the game servers should behave like ETTV does: they will stream the match to the anticheat server.
This server then extracts the player behaviour variables and performs the actual detection. 
I think that with modern hardware it can be even done on-the-fly. Otherwise the game servers can upload the full match demo to the anticheat server that will add it to a queue and perform the detection quetly.

After that, if a player is above the "cheat threshold", the server broadcasts an alert (player, match) to all the admins. We can use a scoring system that keeps track of the player "cheat index" and if multiple positives are detected, manual admin review is needed.
